### PR TITLE
feat(scheduledworkflow): add functionality to relay on pipeline id instead of copy workflow. Fixes #4752

### LINKED ISSUE #4752: [Discussion] Future of recurring run (a.k.a job or scheduled workflow)
There were a lot of discussion in https://github.com/kubeflow/pipelines/issues/3789 that I think worth grouping in a new issue.

> Our conclusion for Job is:
> We may work on a refactoring that swaps out underlying `ScheduledWorkflow` implementation with kubernetes `CronJob`, that will be best for long term maintenance. We don't want to maintain our own `ScheduledWorkflowController` that is a feature subset of `CronJob`. Using `CronJob` as implementation details allow first party integration like UI management, metadata for pipeline runs for CronJob, while keeping maintenance cost low for the controller.
>
> We don't have an ETA for the refactoring right now, because `Job` is not our priority, we'll triage based on future user requirements.

Originally posted in https://github.com/kubeflow/pipelines/issues/3789#issuecomment-662296647

### PR DESCRIPTION
**Pull request by NikeNano**
**Description of your changes:**
Part of https://github.com/kubeflow/pipelines/issues/4752

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: backend/Dockerfile.scheduledworkflow [hunk 1]
```diff
@@ -24,17 +24,17 @@ RUN apk update && apk upgrade && \
 RUN GO111MODULE=on go build -o /bin/controller backend/src/crd/controller/scheduledworkflow/*.go
 # Check licenses and comply with license terms.
 RUN ./hack/install-go-licenses.sh
-RUN go-licenses csv /bin/controller > /tmp/licenses.csv && \
-    diff /tmp/licenses.csv backend/third_party_licenses/swf.csv && \
-    go-licenses save /tmp/licenses.csv --save_path /tmp/NOTICES
+#RUN go-licenses csv /bin/controller > /tmp/licenses.csv && \
+#    diff /tmp/licenses.csv backend/third_party_licenses/swf.csv && \
+#    go-licenses save /tmp/licenses.csv --save_path /tmp/NOTICES
```
Comment (NikeNano) [inline backend/Dockerfile.scheduledworkflow:29]: How do I update the licenses?

File: backend/src/crd/controller/scheduledworkflow/main.go [hunk 2]
```diff
@@ -80,13 +89,31 @@ func main() {
...
+		mlPipelineAPIServerBasePath,
+		mlPipelineAPIServerName,
+		mlPipelineServiceHttpPort,
+		mlPipelineServiceGRPCPort)
+	if err != nil {
+		log.Fatalf("Error creating ML pipeline API Server client: %v", err)
+	}
+
+	//initConnectionTimeout := "6m" //common.GetDurationConfig(initConnectionTimeout)
+	// Here we ingest the objectstorage
```
Comment (NikeNano) [inline backend/src/crd/controller/scheduledworkflow/main.go:104]: Will remove this.

File: backend/src/crd/controller/scheduledworkflow/main.go [hunk 3]
```diff
@@ -122,3 +154,36 @@ func init() {
...
+	secretKey, ok := os.LookupEnv("OBJECTSTORECONFIG_SECRETACCESSKEY")
+	if !ok {
+		log.Info("OBJECTSTORECONFIG_SECRETACCESSKEY is not set")
+	}
+
+	bucketName, ok := os.LookupEnv("OBJECTSTORECONFIG_BUCKETNAME")
+	if !ok {
+		log.Info("OBJECTSTORECONFIG_BUCKETNAME is not set")
+	}
+
```
Comment (NikeNano) [inline backend/src/crd/controller/scheduledworkflow/main.go:182]: Will clean this up and remove the objectstoreconfig.

File: backend/src/crd/controller/scheduledworkflow/util/scheduled_workflow.go [hunk 4]
```diff
@@ -148,6 +158,37 @@ func (s *ScheduledWorkflow) getWorkflowParametersAsMap() map[string]string {
 	return resultAsMap
 }
 
+func (s *ScheduledWorkflow) UpdateWorkflow(
```
Comment (NikeNano) [inline backend/src/crd/controller/scheduledworkflow/util/scheduled_workflow.go:161]: Update NewWorkflow.

**Comment by Bobgy** [comment]
Thank so much for building these!
I am busy with 1.7 release this week, I will start reviewing next week!

: )

**Comment by NikeNano** [comment]
> Thank so much for building these!
> I am busy with 1.7 release this week, I will start reviewing next week!
> 
> : )

Will be on vacation and will take some time off the grid, back on the 23. So no rush :)

**Comment by NikeNano** [comment]
> Thank so much for building these!
> I am busy with 1.7 release this week, I will start reviewing next week!
> 
> : )

I am back :) Would be awesome with some feedback on this @Bobgy

File: backend/src/crd/controller/scheduledworkflow/util/setup.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/pkg/errors"
+
+	api "github.com/kubeflow/pipelines/backend/api/go_client"
+)
+
+const (
+	addressTemp = "%s:%s"
+)
+
+func WaitForAPIAvailable(initializeTimeout time.Duration, basePath string, apiAddress string) error {
```
Comment (Bobgy) [inline backend/src/crd/controller/scheduledworkflow/util/setup.go:19]: nit: this duplicates https://github.com/kubeflow/pipelines/blob/cc83e1089b573256e781ed2e4ac90f604129e769/backend/src/common/util/service.go#L31.

Consider using that function directly?

File: backend/src/crd/controller/scheduledworkflow/controller.go [hunk 6]
```diff
@@ -457,65 +465,107 @@ func (c *Controller) submitNextWorkflowIfNeeded(ctx context.Context, swf *util.S
...
+			swf.Name, workflowName, commonutil.FormatTimeForLogging(nextScheduledEpoch))
+
+		// The pipeline version is not allways set, optional.
+		resourceRefs := []*api.ResourceReference{&api.ResourceReference{
+			Key: &api.ResourceKey{
+				Type: api.ResourceType_PIPELINE,
+				Id:   *swf.Spec.PipelineVersionID,
+			},
+		}}
+		if swf.Spec.PipelineVersionID != nil {
```
Comment (Bobgy) [inline backend/src/crd/controller/scheduledworkflow/controller.go:528]: discussion: can you confirm KFP API behavior, I recall that it's valid to just specify the pipeline version ID, because we can query API to get corresponding pipeline.

Therefore, I think behavior here can be that:
* when version ID is specified, but not pipeline ID, we should still be able to process it.

What do you think? Are there challenges here?

File: backend/src/crd/controller/scheduledworkflow/controller.go [hunk 7]
```diff
@@ -457,65 +465,107 @@ func (c *Controller) submitNextWorkflowIfNeeded(ctx context.Context, swf *util.S
...
+			ScheduledWorkflow: swf.Name,
+			Workflow:          workflowName,
+		}).Infof("PipelineID used when submitting run",
+			swf.Name, workflowName, commonutil.FormatTimeForLogging(nextScheduledEpoch))
+
+		// The pipeline version is not allways set, optional.
+		resourceRefs := []*api.ResourceReference{&api.ResourceReference{
+			Key: &api.ResourceKey{
+				Type: api.ResourceType_PIPELINE,
+				Id:   *swf.Spec.PipelineVersionID,
```
Comment (Bobgy) [inline backend/src/crd/controller/scheduledworkflow/controller.go:525]: nit: typo, we should specify pipeline id

File: backend/src/crd/controller/scheduledworkflow/controller.go [hunk 8]
```diff
@@ -457,65 +465,107 @@ func (c *Controller) submitNextWorkflowIfNeeded(ctx context.Context, swf *util.S
...
+				},
+			})
+		}
+
+		_, err := c.runServiceClient.CreateRun(ctx,
+			&api.CreateRunRequest{Run: &api.Run{
+				Name:               workflowName,
+				ResourceReferences: resourceRefs,
+				PipelineSpec: &api.PipelineSpec{
+					PipelineId: *swf.Spec.PipelineID,
```
Comment (Bobgy) [inline backend/src/crd/controller/scheduledworkflow/controller.go:542]: nit: it's been a while, I recall that specifying pipeline id in the spec is deprecated (but still backward compatible supported), resource refs are the canonical way to specify pipelines/versions

File: backend/src/crd/controller/scheduledworkflow/controller.go [hunk 9]
```diff
@@ -457,65 +465,107 @@ func (c *Controller) submitNextWorkflowIfNeeded(ctx context.Context, swf *util.S
...
+		)
+		if err != nil {
+			return false, "", err
+		}
+
+		return true, workflowName, nil
 
+	}
+
+	newWorkflow, err := swf.NewWorkflow(nextScheduledEpoch, nowEpoch)
```
Comment (Bobgy) [inline backend/src/crd/controller/scheduledworkflow/controller.go:555]: nit: add a comment explaining this is kept for legacy backward compatible support?

File: backend/src/apiserver/resource/resource_manager.go [hunk 10]
```diff
@@ -735,9 +735,16 @@ func (r *ResourceManager) CreateJob(ctx context.Context, apiJob *api.Job) (*mode
 	if err != nil {
 		return nil, util.Wrap(err, "Create job failed")
 	}
+	var pipelineID *string
+	if apiJob.PipelineSpec != nil {
+		if apiJob.PipelineSpec.GetPipelineId() != "" {
+			pipelineID = &apiJob.PipelineSpec.PipelineId
```
Comment (Bobgy) [inline backend/src/apiserver/resource/resource_manager.go:741]: nit: I recall that we can also get pipeline id from resource ref, can you confirm?

File: backend/src/crd/controller/scheduledworkflow/main.go [hunk 11]
```diff
@@ -97,28 +119,24 @@ func main() {
...
-	initEnv()
 
 	flag.StringVar(&kubeconfig, "kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
 	flag.StringVar(&masterURL, "master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
 	flag.StringVar(&namespace, "namespace", "", "The namespace name used for Kubernetes informers to obtain the listers.")
+	flag.StringVar(&mlPipelineAPIServerName, "pipelineAPIServerName", "ml-pipeline", "Name of the ML pipeline API server.")
+	flag.StringVar(&mlPipelineAPIServerBasePath, "pipelineAPIServerBasePath",
+		"/apis/v1beta1", "The base path for the ML pipeline API server.")
+	flag.StringVar(&mlPipelineServiceHttpPort, "pipelineServiceHttpPort", "8888", "Http Port of the ML pipeline API server.")
+	flag.StringVar(&mlPipelineServiceGRPCPort, "pipelineServiceGRPCPort", "8887", "GRPC Port of the ML pipeline API server.")
```
Comment (Bobgy) [inline backend/src/crd/controller/scheduledworkflow/main.go:137]: will we be able to get these values from env vars like `ML_PIPELINE_SERVICE_HOST` and `ML_PIPELINE_SERVICE_PORT`?

**Review by Bobgy — COMMENTED** [review]
Thanks a lot! I did a thorough review

File: backend/src/apiserver/resource/resource_manager.go (hunk 10 above)
Comment (NikeNano) [reply backend/src/apiserver/resource/resource_manager.go:741]: I dont think we can do that: 

https://github.com/kubeflow/pipelines/blob/d03277fe4ed79f2bc4780ca1cdb6bab9469eeb61/backend/src/apiserver/resource/resource_manager_util.go#L248-L250

File: backend/src/apiserver/resource/resource_manager.go [hunk 12]
```diff
@@ -735,9 +735,16 @@ func (r *ResourceManager) CreateJob(ctx context.Context, apiJob *api.Job) (*mode
 	if err != nil {
 		return nil, util.Wrap(err, "Create job failed")
 	}
+	var pipelineID *string
+	if apiJob.PipelineSpec != nil {
```
[the remark this answers is not in the document]
Comment (NikeNano) [reply backend/src/apiserver/resource/resource_manager.go:739]: Great!

**Comment by NikeNano** [comment]
Will continue to fix it tomorrow.

File: backend/src/crd/controller/scheduledworkflow/main.go (hunk 11 above)
Comment (NikeNano) [reply backend/src/crd/controller/scheduledworkflow/main.go:136]: How would `ML_PIPELINE_SERVICE_HOST ` replace `mlPipelineAPIServerBasePath ` and `mlPipelineAPIServerName` I might be missing something here?

**Comment by NikeNano** [comment]
Fail to recreate the failing `kubeflow-pipeline-backend-test` locally atm.

File: backend/src/crd/controller/scheduledworkflow/controller.go (hunk 6 above)
Comment (Bobgy) [reply backend/src/crd/controller/scheduledworkflow/controller.go:523]: Makes sense!

File: backend/src/apiserver/resource/resource_manager.go (hunk 10 above)
Comment (Bobgy) [reply backend/src/apiserver/resource/resource_manager.go]: Thanks for confirming! You are right

File: backend/src/crd/controller/scheduledworkflow/controller.go (hunk 8 above)
Comment (Bobgy) [reply backend/src/crd/controller/scheduledworkflow/controller.go]: Thanks for confirming! You are right

File: backend/src/crd/controller/scheduledworkflow/main.go (hunk 11 above)
Comment (Bobgy) [reply backend/src/crd/controller/scheduledworkflow/main.go:136]: I see, since we already have the util, let's keep using it.

**Comment by NikeNano** [comment]
I fail to reproduce the failing test: 

```bash
docker run -it golang:1.13.15-stretch /bin/bash
git clone --single-branch --branc nikenano/UpdateScheuleRunController  https://github.com/NikeNano/pipelines
cd pipelines/backend/src/crd/controller/scheduledworkflow/util/
go test -v
```

Do you have some ideas @Bobgy? Thank you

**Comment by NikeNano** [comment]
I fail to recreate the failed test: 

```
docker run -it golang:1.13.15-stretch /bin/bash
git clone --single-branch --branch nikenano/UpdateScheuleRunController  https://github.com/NikeNano/pipelines.git
cd pipelines/backend/src/crd/controller/scheduledworkflow/util/
go test -v
```

Maybe you have some ideas @Bobgy? Thanks!

**Comment by Bobgy** [comment]
The test runs based on merging the PR to master. Could that be the cause?

**Comment by NikeNano** [comment]
> The test runs based on merging the PR to master. Could that be the cause?

Ahh will rebase and test it!

**Comment by NikeNano** [comment]
Still fail to figure out where it comes from :(

**Comment by Bobgy** [comment]
/assign @capri-xiyue
Can you help NikeNano troubleshoot why the backend test is failing?
This PR is an important step for enabling recurring runs for KFP v2.

File: backend/src/crd/controller/scheduledworkflow/controller.go [hunk 13]
```diff
@@ -457,65 +465,111 @@ func (c *Controller) submitNextWorkflowIfNeeded(ctx context.Context, swf *util.S
...
 
-	// If the workflow is not found, we need to create it.
-	newWorkflow, err := swf.NewWorkflow(nextScheduledEpoch, nowEpoch)
-	createdWorkflow, err := c.workflowClient.Create(ctx, swf.Namespace, newWorkflow)
+	if swf.Spec.PipelineID != nil {
+
+		log.WithFields(log.Fields{
+			ScheduledWorkflow: swf.Name,
+			Workflow:          workflowName,
+		}).Infof("PipelineID used when submitting run",
```
Comment (capri-xiyue) [inline backend/src/crd/controller/scheduledworkflow/controller.go:518]: @NikeNano  I think this is the root cause of the backend test failure. You didn't add formatting directive like "%v". You should be able to reproduce it when you run the backend test in local.

**Comment by capri-xiyue** [comment]
> /assign @capri-xiyue
> Can you help NikeNano troubleshoot why the backend test is failing?
> This PR is an important step for enabling recurring runs for KFP v2.

See https://github.com/kubeflow/pipelines/pull/6207#discussion_r706371570

File: backend/src/crd/controller/scheduledworkflow/controller.go (hunk 13 above)
Comment (NikeNano) [reply backend/src/crd/controller/scheduledworkflow/controller.go:518]: Thank you @capri-xiyue approciate, this is obviously wrong but what did you run to catch it? For me, locally the tests seems to run fine, I cant figure out what I am missing ...

Comment (capri-xiyue) [reply backend/src/crd/controller/scheduledworkflow/controller.go]: I use the IDE `GOLAND` to run the backend tests, looks like `GOLAND` uses command `go test backend`.
Did you run the tests in local with go command line directly in terminal? I found the output of the go command line in terminal is not easy to read, sometimes I  ignore some errors. For me, reading the results of running tests in IDE  is more user friendly.

Comment (NikeNano) [reply backend/src/crd/controller/scheduledworkflow/controller.go]: Ok, thank you. I use the terminal to run the test, agree that it is really hard to read the outputs and scan for the failed test. I looked in the raw logs and could not find anything in them either. Will give `GOLAND` a test :)

**Comment by NikeNano** [comment]
UPDATE: Working on understanding the persistance agent and how the scheduled runs actually are updated in the database. Will update when I am ready for a review and found what needs to be fixed.

**Comment by Bobgy** [comment]
Hi @NikeNano, to avoid expanding size of the PR too much, I'd recommend implementing the minimal fixes you need in persistence agent to get tests passing. You can add full persistence agent support in a separate PR.

**Comment by NikeNano** [comment]
Could you take a look again @Bobgy, thanks.

**Review by Bobgy — COMMENTED** [review]
Some early comments first, will continue to review : )

File: backend/api/run.proto [hunk 14]
```diff
@@ -262,8 +262,22 @@ message Run {
   // Output. The metrics of the run. The metrics are reported by ReportMetrics
   // API.
   repeated RunMetric metrics = 9;
+
+  // Optional input field. Specify which resource is the owner.
+  // This is used for scheduled runs. 
+  Owner Owner = 15;
```
[the remark this answers is not in the document]
Comment (NikeNano) [reply backend/api/run.proto:268]: Could you elaborate don't fully follow what you suggest here. Thanks.

**Comment by NikeNano** [comment]
@Bobgy  do you have the possibility to take a look again? Thanks

**Comment by NikeNano** [comment]
@Bobgy and @capri-xiyue would be awesome if you guys could come with feedback on this in order to progress with it, thank you.

File: backend/src/apiserver/model/pipeline_spec.go [hunk 15]
```diff
@@ -19,6 +19,10 @@ type PipelineSpec struct {
 	// a pipeline ID.
 	PipelineId string `gorm:"column:PipelineId; not null"`
 
+	// PipelineVersionID will be optional. It's available only if the resource is created through
+	// a pipeline version ID.
+	PipelineVersionId string `gorm:"column:PipelineVersionId; not null"`
```
[the remark this answers is not in the document]
Comment (capri-xiyue) [reply backend/src/apiserver/model/pipeline_spec.go:24]: I checked the code. This field is not exposed to users, user still need to specify PipelineVersionId through Resource References, right?

File: backend/api/run.proto (hunk 14 above)
[the remark this answers is not in the document]
Comment (capri-xiyue) [reply backend/api/run.proto:268]: @Bobgy Can you further elaborate this?

**Comment by capri-xiyue** [comment]
@NikeNano  Can you help clarify why we need this change?

**Comment by NikeNano** [comment]
> @NikeNano Can you help clarify why we need this change?

As in the whole PR?

File: backend/src/apiserver/model/pipeline_spec.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (NikeNano) [reply backend/src/apiserver/model/pipeline_spec.go:24]: I think you are correct here @capri-xiyue and that this is a misstake on my side, I think we can drop this field in the implementation and leave it to future work to replace the resource references.  Will update tomorrow.

**Comment by Bobgy** [comment]
Sorry, a bunch of things kept me busy. I am resuming on review now. : )

File: backend/api/run.proto (hunk 14 above)
[the remark this answers is not in the document]
Comment (NikeNano) [reply backend/api/run.proto:268]: As I understand `api.ResourceReference` and `model.ResourceReference` depend on different implementations and  there are no `api.ResourceType` for ownership. However to stay consistent I suggest we added it. Could you confirm @Bobgy ?

**Comment by NikeNano** [comment]
working on it, but the rebase required more work then I expected.

**Comment by NikeNano** [comment]
> Hi @NikeNano, have you considered?
> 
> 1. removing some unnecessary changes from your PR (like the proto changes)
> 2. squash your own commits to one commit
> 3. rebase to master and resolve conflicts
> 
> In this way, you won't need to resolve conflicts per commit, so it may save some work.

Thanks, I had to make some updates to handle the V2 changes which where the main issues. Have had a lot a work as well so took longer to fix then I expected.

**Comment by Bobgy** [comment]
Thank you for the updates! I'll take a look today or tomorrow.

File: backend/Dockerfile.scheduledworkflow [hunk 16]
```diff
@@ -24,7 +24,7 @@ RUN apk update && apk upgrade && \
 RUN GO111MODULE=on go build -o /bin/controller backend/src/crd/controller/scheduledworkflow/*.go
 # Check licenses and comply with license terms.
 RUN ./hack/install-go-licenses.sh
-RUN go-licenses csv /bin/controller > /tmp/licenses.csv && \
+RUN go-licenses csv /bin/controller > /tmp/licenses.csv  && \
```
Comment (Bobgy) [inline backend/Dockerfile.scheduledworkflow:27]: nit: revert unnecessary changes?

File: backend/src/apiserver/resource/resource_manager.go [hunk 17]
```diff
@@ -369,11 +375,56 @@ func (r *ResourceManager) CreateRun(ctx context.Context, apiRun *api.Run) (*mode
...
+		// Disable istio sidecar injection if not specified
+		workflow.SetAnnotationsToAllTemplatesIfKeyNotExist(util.AnnotationKeyIstioSidecarInject, util.AnnotationValueIstioSidecarInjectDisabled)
+		// Add a KFP specific label for cache service filtering. The cache_enabled flag here is a global control for whether cache server will
+		// receive targeting pods. Since cache server only receives pods in step level, the resource manager here will set this global label flag
+		// on every single step/pod so the cache server can understand.
+		// TODO: Add run_level flag with similar logic by reading flag value from create_run api.
+		workflow.SetLabelsToAllTemplates(util.LabelKeyCacheEnabled, common.IsCacheEnabled())
+
+		err = OverrideParameterWithSystemDefault(workflow, apiRun)
+	*/
```
Comment (Bobgy) [inline backend/src/apiserver/resource/resource_manager.go:403]: what's the plan about these lines?

File: backend/src/apiserver/resource/resource_manager.go [hunk 18]
```diff
@@ -410,6 +461,23 @@ func (r *ResourceManager) CreateRun(ctx context.Context, apiRun *api.Run) (*mode
...
 }
 
+type owner struct {
+	name string
+	id   string
+}
+
+func getOwner(resourceRef []*api.ResourceReference) owner {
+	for _, ref := range resourceRef {
+		if ref.Relationship == api.Relationship_OWNER {
```
Comment (Bobgy) [inline backend/src/apiserver/resource/resource_manager.go:471]: nit: this can also mean an owner in KFP resource relationship. I think for adding Kubernetes owners, we should only add it for "runs owned by jobs".

File: backend/src/apiserver/resource/resource_manager_test.go [hunk 19]
```diff
@@ -18,11 +18,12 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
-	"github.com/kubeflow/pipelines/backend/src/apiserver/template"
 	"strings"
 	"testing"
 	"time"
 
+	"github.com/kubeflow/pipelines/backend/src/apiserver/template"
```
Comment (Bobgy) [inline backend/src/apiserver/resource/resource_manager_test.go:25]: nit: revert formatting changes unrelated to this PR?

We should probably format all the code again and enforce it in presubmit tests, but for this PR, let's not include them, so it's easier to review. (However, to save your efforts, for any file that you modified, I'm OK with keeping formatting changes.)

File: backend/src/crd/pkg/apis/scheduledworkflow/v1beta1/types.go [hunk 20]
```diff
@@ -76,6 +76,18 @@ type ScheduledWorkflowSpec struct {
...
+	// +optional
+	PipelineID *string `json:"pipelineID,omitempty"`
+
+	// The pipeline version id
+	// +optional
+	PipelineVersionID *string `json:"pipelineVersionID,omitempty"`
+
+	// The experiment id
+	// +optional
+	ExperimentID *string `json:"experimentID,omitempty"`
```
Comment (Bobgy) [inline backend/src/crd/pkg/apis/scheduledworkflow/v1beta1/types.go:89]: What's your choice for adding pipeline parameters & labels here?
As discussed in https://github.com/kubeflow/pipelines/pull/6207#issuecomment-953256544.

We can have a GVC chat etc if you prefer.

**Review by Bobgy — COMMENTED** [review]
Some comments first, still reading through the whole PR.

File: backend/src/crd/controller/scheduledworkflow/controller.go [hunk 21]
```diff
@@ -457,65 +465,138 @@ func (c *Controller) submitNextWorkflowIfNeeded(ctx context.Context, swf *util.S
...
+		}
+
+		_, err = c.workflowClient.Create(ctx, swf.Namespace, newWorkflow)
+		return true, workflowName, err
+
+	}
+
+	formatter := commonutil.NewSWFParameterFormatter("", nextScheduledEpoch, nowEpoch, swf.NextIndex())
+	formattedParams := formatter.FormatWorkflowParameters(swf.GetWorkflowParametersAsMap())
+	parameters := OverrideParameters(swf.Spec.Workflow.Spec.Arguments.Parameters, formattedParams)
```
Comment (Bobgy) [inline backend/src/crd/controller/scheduledworkflow/controller.go:535]: Note, parameters will be a different format now for v2.
https://github.com/kubeflow/pipelines/blob/60233f2a1da9e206ab0077193f091a8d39dcc05a/backend/api/pipeline_spec.proto#L47

That's why I mentioned adding a field in swf CRD that exactly mimics the PipelineSpec object.

Curious about your thoughts

**Review by Bobgy — COMMENTED** [review]
Finished review

**Comment by Bobgy** [comment]
Great work! I think you might not be fully understanding the changes for v2. Feel free to ask more questions if you have anything unclear.

**Comment by NikeNano** [comment]
Will pick this up today and continue the work.

File: backend/src/apiserver/resource/resource_manager_test.go (hunk 19 above)
Comment (NikeNano) [reply backend/src/apiserver/resource/resource_manager_test.go:25]: I will clean it up,  no worries.

File: backend/src/apiserver/resource/resource_manager.go (hunk 17 above)
Comment (NikeNano) [reply backend/src/apiserver/resource/resource_manager.go:403]: Should be removed.

File: backend/src/crd/controller/scheduledworkflow/controller.go (hunk 21 above)
Comment (NikeNano) [reply backend/src/crd/controller/scheduledworkflow/controller.go:535]: See your point and agree that we should have it as a seperate field with a map[string]string.

File: backend/src/apiserver/template/argo_template.go [hunk 22]
```diff
@@ -89,15 +90,38 @@ func (t *Argo) ScheduledWorkflow(apiJob *api.Job) (*scheduledworkflow.ScheduledW
 	if err != nil {
 		return nil, util.Wrap(err, "Create job failed")
 	}
+	//
+
+	var pipelineID *string
```
[the remark this answers is not in the document]
Comment (NikeNano) [reply backend/src/apiserver/template/argo_template.go:95]: I see your point on this, I will update it will require some refactoring. Will continue to push it tomorrow and hopefully have something up then. Will put prio on this during the week so hope I can get some time to push it again.

[the remark this answers is not in the document]
Comment (NikeNano) [reply backend/src/apiserver/template/argo_template.go:95]: I guess we could add the raw manifest incase the pipelineID and pipelineVersionID is empty as an alternative.

File: backend/src/crd/pkg/apis/scheduledworkflow/v1beta1/types.go (hunk 20 above)
Comment (NikeNano) [reply backend/src/crd/pkg/apis/scheduledworkflow/v1beta1/types.go:89]: Yes, will add it, missed it. Have been jumping around a bit to much between things.

File: backend/src/apiserver/template/argo_template.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (NikeNano) [reply backend/src/apiserver/template/argo_template.go:95]: @Bobgy

[the remark this answers is not in the document]
Comment (Bobgy) [reply backend/src/apiserver/template/argo_template.go:95]: Agree, we can store the raw manifest (v2 pipeline spec) into swf CR, similar to how users can create a run using raw manifest.

File: backend/src/crd/pkg/apis/scheduledworkflow/v1beta1/types.go (hunk 20 above)
Comment (Bobgy) [reply backend/src/crd/pkg/apis/scheduledworkflow/v1beta1/types.go:89]: No worries, thank you for working on this!

File: backend/src/crd/controller/scheduledworkflow/controller.go [hunk 23]
```diff
@@ -414,19 +422,19 @@ func (c *Controller) syncHandler(ctx context.Context, key string) (
 			wraperror.Wrapf(err, "Syncing ScheduledWorkflow (%v): transient failure, can't fetch completed workflows: %v", name, err)
 	}
 
-	workflow, nextScheduledEpoch, err := c.submitNextWorkflowIfNeeded(ctx, swf, len(active), nowEpoch)
+	submitted, nextScheduledEpoch, err := c.submitNextWorkflowIfNeeded(ctx, swf, len(active), nowEpoch)
```
Comment (Bobgy) [inline backend/src/crd/controller/scheduledworkflow/controller.go:425]: example change that can be extracted out

**Review by Bobgy — COMMENTED** [review]
I'm not sure how big an ask this is, but I can see that we will need to discuss the design with v2 in context, so there might be more design discussions.

I wonder which changes are those we can separate out into small PRs we can merge first. For example, I think refactoring return value of submitNextWorkflowIfNeeded to bool can totally be a small PR we can accept.

In this way, we can move faster and review more effectively.

File: backend/src/crd/controller/scheduledworkflow/controller.go (hunk 23 above)
Comment (NikeNano) [reply backend/src/crd/controller/scheduledworkflow/controller.go:425]: > I'm not sure how big an ask this is, but I can see that we will need to discuss the design with v2 in context, so there might be more design discussions.
> 

As long as it bring us forward, I am fine with it. I will leave this one up for now but start to bring things out. Will start to take a look at your suggestion of the `submitNextWorkflowIfNeeded `.

Comment (NikeNano) [reply backend/src/crd/controller/scheduledworkflow/controller.go:425]: @Bobgy : https://github.com/kubeflow/pipelines/pull/6918

**Comment by NikeNano** [comment]
I will break out and try to remove the usage of `Template.ScheduledWorkflow` as a next step to work toward the end goal.

**Comment by Matts966** [comment]
Hi, this PR is very desirable for our usage paths!
Current recurring run has some problems such as

- `RUN_ID_PLACE_HOLDER` is not updated when cloning runs published by recurring run.
- We can not use latest pipeline version in runs published by recurring run

Can this PR be more prioritized? Thank you :bow:

**Comment by NikeNano** [comment]
Is this a feature we would like to push forward @zijianjoy I have not had time before. It would be great to know there is bandwith for review and that we like to have this feature before I spend time on it.

**Comment by Tomcli** [comment]
Hi, our team is working on supporting KFP v2 to run on Tekton. This PR can really help us generalize the spec for scheduled workflows in V2. My teammate @yhwang changed the scheduled workflows spec type to make it work for our use case for now but we are hoping to see this PR can move forward. Is there any chance our team can help on making this PR merge into KFP?

**Comment by NikeNano** [comment]
> Hi, our team is working on supporting KFP v2 to run on Tekton. This PR can really help us generalize the spec for scheduled workflows in V2. My teammate @yhwang changed the scheduled workflows spec type to make it work for our use case for now but we are hoping to see this PR can move forward. Is there any chance our team can help on making this PR merge into KFP?

I am currently don't have the time to work on this so feel free to take it over if you like to @Tomcli!

**Comment by NikeNano** [comment]
> Is this a feature we would like to push forward @zijianjoy I have not had time before. It would be great to know there is bandwith for review and that we like to have this feature before I spend time on it.

@Tomcli  i do remember getting feedback as one of the biggest issues though and sadly also partly why I stopped contributing(also lack of time).
